Source/Core/VideoBackends/Vulkan/VKTexture.cpp

```diff
@@ -794,6 +892,55 @@ void VKStagingTexture::CopyFromTexture(const AbstractTexture* src,
   m_flush_fence_counter = g_command_buffer_mgr->GetCurrentFenceCounter();
 }
 
+void VKStagingTexture::CopyFromTextureToLinearImage(const VKTexture* src_tex,
+                                                    const MathUtil::Rectangle<int>& src_rect,
+                                                    u32 src_layer, u32 src_level,
+                                                    const MathUtil::Rectangle<int>& dst_rect)
+{
+  // The proprietary Qualcomm driver allocates a temporary image when copying from an image
+  // with optimal tiling (VK_IMAGE_TILING_OPTIMAL) to a buffer.
+  // That allocation is very slow, so we just do it ourself and reuse the intermediate image.
+
+  const VkImageAspectFlags aspect = VKTexture::GetImageViewAspectForFormat(src_tex->GetFormat());
+
+  VkImageMemoryBarrier linear_image_barrier = {};
+  linear_image_barrier.sType = VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER;
+  linear_image_barrier.pNext = nullptr;
+  linear_image_barrier.srcAccessMask = 0;
+  linear_image_barrier.dstAccessMask = VK_ACCESS_TRANSFER_WRITE_BIT | VK_ACCESS_TRANSFER_READ_BIT;
+  linear_image_barrier.oldLayout = VK_IMAGE_LAYOUT_UNDEFINED;
+  linear_image_barrier.newLayout = VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL;
+  linear_image_barrier.image = m_linear_image;
+  linear_image_barrier.subresourceRange = {aspect, 0, 1, 0, 1};
+  vkCmdPipelineBarrier(g_command_buffer_mgr->GetCurrentCommandBuffer(),
+                       VK_PIPELINE_STAGE_TRANSFER_BIT, VK_PIPELINE_STAGE_TRANSFER_BIT, 0, 0,
+                       nullptr, 0, nullptr, 1, &linear_image_barrier);
+
+  VkImageBlit blit;
+  blit.srcSubresource = {aspect, src_level, src_layer, 1};
+  blit.dstSubresource.layerCount = 1;
+  blit.dstSubresource.baseArrayLayer = 0;
+  blit.dstSubresource.mipLevel = 0;
+  blit.dstSubresource.aspectMask = linear_image_barrier.subresourceRange.aspectMask;
+  blit.srcOffsets[0] = {src_rect.left, src_rect.top, 0};
+  blit.srcOffsets[1] = {static_cast<s32>(blit.srcOffsets[0].x + src_rect.GetWidth()),
+                        static_cast<s32>(blit.srcOffsets[0].y + src_rect.GetHeight()), 1};
+  blit.dstOffsets[0] = {0, 0, 0};
+  blit.dstOffsets[1] = {dst_rect.GetWidth(), dst_rect.GetHeight(), 1u};
+
+  vkCmdBlitImage(g_command_buffer_mgr->GetCurrentCommandBuffer(), src_tex->GetImage(),
+                 VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL, m_linear_image,
+                 VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL, 1, &blit, VK_FILTER_LINEAR);
```

Comment:
Is `VK_FILTER_LINEAR` desired? `CopyFromTexture` asserts that `src_rect.GetWidth() == dst_rect.GetWidth() && src_rect.GetHeight() == dst_rect.GetHeight()`, so this shouldn't be doing any scaling, in which case I think `VK_FILTER_LINEAR` will just introduce blurring and (slightly) worse performance compared to `VK_FILTER_NEAREST`.